L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,296 @@
+* Load balancing strategies. Includes built-in support for pick first and round-robin load balancing. Can create custom load balancers.
+* Connectivity status at sub-channel and channel level
+* Wait for ready
+
+## Proposal
+
+![Diagram](L80_graphics/diagram.png)
+
+### Initialization
+
+Channels are created with:
+
+* An address
+* An optional service provider
+* An optional service config
+
+```csharp
+var channel = GrpcChannel.ForAddress(
+    "dns:///backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Resolver is created from address scheme.
+  * Existing logic allows for `http` and `https`. These schemes are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme, and it will create a resolver. For example, `dns` will match to `DnsAddressResolverFactory`, creating a `DnsAddressResolver`.
+* Load balancer is created from the load balancer name in service config.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified, then `pick_first` is used), creating a load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. The controller calls back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` starts the resolver and prompts it to run immediately.
+* Resolver pushes addresses and optional service config to `ConnectionManager`.
+* Load balancer is created.
+  * Resolver can return a service config. If `DisableServiceConfig` is not `true` then the resolver's service config is used to get load balancer policy.
+  * Channel service config option is used as a fallback.
+  * When no load balancer policy has been specified then a `PickFirstLoadBalancer` is created.
+* Balancer uses `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel starts connecting when `SubChannel.RequestConnection()` is called. Sub-channels shut down when disposed.
+* Balancer is notified of changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker is empty, or the picker has an error then `PickAsync` will wait for a picker with a ready sub-channel if `WaitForReady` has been set. If `WaitForReady` is not set, then the error will be thrown back to the caller.
```

Comment:
I don't really understand what "the picker is empty" means.  Do you mean that the picker returns FAIL?

Reply:
Yes. I'll update the doc to use these common terms.